### ISSUE TITLE
Controlling fine-tuning in more detail?

### ISSUE THREAD

**Issue by Alex-Zien** [deepset-ai/FARM#141]
(How) Can I control the schedule of fine-tuning the language model's parameters?

More specifically, I'd like to first just train the PredictionHead on top of a frozen LM, then after a while allow for a slowly increasing learning rate for the LM (while continuing a adjust the PredictionHead with a fixed lr).

Or I'd like to perform layer-wise fine-tuning: Start training with the LM completely frozen, the unfreeze the top layer and train for a while, the unfreeze the next layer, and so on.

Can this be made possible with FARM?

**Comment by tholor** [comment]
Hi @Alex-Zien, 

I agree that this could be helpful and we definitely have it on our roadmap for FARM. However, we have some other features that we will work on first (e.g. caching datasets, improving speed, more LMs ...). So if this is something you really need urgently, I would encourage you to implement this by yourself and to raise a PR. We are happy to support you along the way!

As a starting point: Freezing the full LM should be rather straight forward. You basically need to set `requires_grad=False` for all parameters in the LM:

```
for param in model.parameters():
    param.requires_grad = False 
```
See: 
- https://discuss.pytorch.org/t/how-the-pytorch-freeze-network-in-some-layers-only-the-rest-of-the-training/7088/16
- https://discuss.pytorch.org/t/correct-way-to-freeze-layers/26714

We already have a placeholder method for this in the LM class that you could probably use for this:
https://github.com/deepset-ai/FARM/blob/30aa9209c5b8840e5c31378d8dd2bcad1365c752/farm/modeling/language_model.py#L107-L110

Gradually unfreezing more and more layers during the course of training will require some bigger, additional changes. I can envision different options to implement this: 
1) Adjust the `Trainer` to unfreeze additional layers every x steps in the Trainer
2) Use the LRSchedule that we will introduce with #133 to define different LR for certain parameter groups over time (LR=0 for most layers in the beginning and gradually increasing over time)
3) ...
 
Let me know if that's something you would like to work on!

**Comment by egorgorn** [comment]
@tholor Sorry if this is a dumb question from both a GitHub and BERT newbie, but I think it is related to what @Alex-Zien asked. I guess this is the proper place to request the authors their advice on this package.
My question is much simpler though. When I conduct fine-tuning for a downstream task (which I distinguish from 'fine-tuning LM model with unlabeled in-domain data' aka 'further LM pre-training'), is there a way to control which layer (out of 12 in BERT's case) is fine-tuned and which one is fixed? And what is the default option when I fine-tune for a classification task: Is it to fine-tune all layers or is it to fine-tune layer 12 only?

**Comment by tholor** [comment]
@egorgorn Sorry for the late reply. I somehow missed your comment here. 

> is there a way to control which layer (out of 12 in BERT's case) is fine-tuned and which one is fixed?

Currently, there's no option to freeze/fix some layers

> And what is the default option when I fine-tune for a classification task: Is it to fine-tune all layers or is it to fine-tune layer 12 only?

The default is to fine-tune all 12 layers (plus the prediction head)

Hope this helps :)
